### ISSUE TITLE
[css-grid-3][masonry] Masonry Syntax Debate

### ISSUE THREAD

**Issue by fantasai** [w3c/csswg-drafts#11243]
The CSSWG's masonry layout draft includes [two competing syntax proposals](https://www.w3.org/TR/css-grid-3/#alternative-proposals) for the feature. There has been a lot of discussion between the two, and there's an [overview of the debate](https://github.com/w3ctag/design-reviews/issues/1003#issuecomment-2456252511) in the TAG review issue.

Opening this issue to focus on the syntax question, since the [previous issue](https://github.com/w3c/csswg-drafts/issues/9041) debating two possible models started with focus on the performance implications of mixed track sizing, etc. that are no longer relevant.

Key links (primary in bold):
- [TAG Review Request](https://github.com/w3ctag/design-reviews/issues/1003): **[Overview of the syntax debate](https://github.com/w3ctag/design-reviews/issues/1003#issuecomment-2456252511)** and **[TAG Response](https://github.com/w3ctag/design-reviews/issues/1003#issuecomment-2489688718)**
- **[Chrome Blog: How should we define CSS masonry?](https://developer.chrome.com/blog/masonry-syntax)**
- **[WebKit Blog: Help us choose the final syntax for Masonry in CSS](https://webkit.org/blog/16026/css-masonry-syntax/)**
- [Ahmad Shadeed: Should masonry be part of grid?](https://ishadeed.com/article/css-grid-masonry/) 
- [Rachel Andrew: Masonry and good defaults](https://rachelandrew.co.uk/archives/2024/09/21/masonry-and-good-defaults/) 
- [Keith J. Grant: Weighing in on CSS masonry](https://keithjgrant.com/posts/2024/05/weighing-in-on-css-masonry/) [published before the merged FPWD]
- [Chrome Blog Comments](https://github.com/w3c/csswg-drafts/issues/9041#ref-issue-1482543319) Chrome directed developers to tack their comments onto the end of the "Alternative Masonry Path" issue, which had been opened to debate a number of things including syntax...
- [WebKit Blog Comments](https://github.com/w3c/csswg-drafts/issues/11060) WebKit opened this fresh issue for syntax-focused comments after the latest developments, closing out an [older issue](https://github.com/w3c/csswg-drafts/issues/10233) from an earlier blog post that collected more general feedback on desired masonry layout capabilities...

**See [summary of current status in this comment below](https://github.com/w3c/csswg-drafts/issues/11243#issuecomment-2762201581)**

**Comment by astearns** [comment]
A couple of questions for folks to (maybe?) spark some discussion:

- What is the strongest argument for the option you do not prefer?
- If you have changed your opinion from one of the options to the other, what was the reason for the change?

**Comment by astearns** [comment]
TAG feedback: https://github.com/w3ctag/design-reviews/issues/1003#issuecomment-2489688718

**Comment by nathanchase** [comment]
FWIW: I still agree that in isolation, `display: masonry` makes more sense due to its brevity, but a lot of the points brought up by @jensimmons (https://webkit.org/blog/16026/css-masonry-syntax/) and @shadeed (https://ishadeed.com/article/css-grid-masonry/) convinced me that the overall context shift is higher to create new keywords, recall syntax differences, and set up default values that could potentially have highly unintended results. Particularly, this matrix is compelling:
![image](https://github.com/user-attachments/assets/567a4219-721b-4c57-906c-a06146078476)

With that in mind, and with the also welcome shift away from using "masonry" as a term and instead leaning on the existing "collapse" keyword already found elsewhere makes a lot more sense for the future.

**Comment by kizu** [comment]
If we were to choose from the two options presented, I prefer the “Just use Grid” option.

I have a bigger take on this, which I just can't find an opportunity to write up properly, but the gist is that with masonry we're mixing two things: [one-dimensional layouts](https://drafts.csswg.org/css-grid-3/#collapse) (without the masonry specifically, as a wider feature), and the masonry auto-placement algorithm. I'd want the first feature as a part of a CSS grid, regardless of the masonry. Then, having one-dimensional grids, introducing an alternative placement algorithm in addition sounds like a reasonable iterative approach.

I am against the `display: masonry` for the reasons presented by Jen and Ahmad, specifically the fallbacks and conditional adjustments aspects. It is much easier to adjust or create a fallback for something when you don't have to change N different properties at the same time. I would prefer learning a slightly different syntax for specific cases rather than an additional set of slightly different properties.

> What is the strongest argument for the option you do not prefer?

I agree with the arguments for the grid-independent option about good defaults and shorthands. However, I can see a future where, instead of putting the current flawed grid aside and giving up on it in favor of a new system, we could see if we could improve what we have. With new keywords added to the mix, they allow us to change the way shorthands behave, and design things better. It is true that authors rarely use the `grid` shorthand, and that means that we should look into how we could improve it for all the cases.

**Comment by astearns** [comment]
Slides from Jen Simmons: https://lists.w3.org/Archives/Public/www-archive/2024Dec/att-0002/Masonry_presentation_to_CSSWG_____Dec_4_2024.pdf
Slides from Alison Maher: https://lists.w3.org/Archives/Public/www-archive/2024Dec/att-0003/display_masonry.pdf

**Comment by JoshTumath** [comment]
Now that I've had more time to consider the opinions raised in our call on Wednesday – and particularly the TAG review which @LeaVerou shared – my mind has changed and I have swayed towards the 'Just use grid' option. I'll explain why.

https://github.com/w3ctag/design-reviews/issues/1003#issuecomment-2489688718

The TAG viewpoint set us a new north star to aim for where _all_ layout methods that do a '2D arrangement of children' become more unified and easier for authors to declare:

> Overall, we think masonry, grid, and wrapping-flexbox should be incorporated into a unified set of properties.

This argument was particularly compelling to me (emphasis added):

> 4. It was argued that the grid-independent syntax makes things easier if the properties need to [diverge in the future](https://developer.chrome.com/blog/masonry-syntax#other_considerations). We think **that is a drawback of that syntax, not a feature.** Even if the WG selects a syntax like the grid-independent one, **it should work hard to ensure that the properties diverge as little as possible in the future,** even if some masonry properties initially appear not to have an obvious behavior for other display modes. It would be even worse to evolve subtle variations in patterns that otherwise appear to match, for example by forgetting to apply a new grid keyword to the parallel masonry property.

This argument trumped my main concerns around how we might want the syntax to diverge. The argument also supported [the vision that the WebKit team gave us at the end of their blog post about a possible future where we could even somehow use the grid properties on flow layouts.](https://webkit.org/blog/16026/css-masonry-syntax/#a-possible-future-feature)

And then the TAG review goes on to talk more broadly about layout syntax.

> We urge the WG to explore ways to unify these so that authors can port more knowledge from one to the other (even if they are implemented as separate code paths internally). For example `grid-auto-flow`, `flex-direction`, `masonry-direction`, and `masonry-fill` all control the same stylistic aspect from an author pov and it seems that they could be more unified to some degree.
> [...]
> These patterns blur the line we've been assuming between 1-d and 2-d layout, and the WG may need to solve the problem by leaning into a continuum between those two options. It would be ideal to find a set of properties that can be used to build all of these layout patterns, instead of [potentially overfitting](https://medium.com/design-bootcamp/overfitting-and-the-problem-with-use-cases-337d9f4bf4d7) to just masonry layouts.

## In conclusion

On behalf of the BBC, I am happy to support the "Just use grid" syntax to move the proposal forward.

However, the TAG review has set out a wider vision for shorthands or other more generic syntaxes for defining layouts. Would it be helpful, as part of this issue, to start blue-skying those, just to help the debate along? I would be interested to see any proposals.

For example, the fact that we can now use `gap` in flex layouts and `align-content` in grid, flex _and now flow_ layouts has been a game changer. It would be great to see how some of the `grid-*` properties could be made more generic. Like @kizu said, 'It is true that authors rarely use the grid shorthand, and that means that we should look into how we could improve it for all the cases.'

**Comment by Monknow** [comment]
I am an outsider here, but wanted to give my grain of sand from an author's point of view. I don't want to get into which proposal will be easier to define in the spec or implement in browsers since that's something that most developers aren't aware of but rather boil it down to what syntax developers will find easier to learn and use, and if the future gives us more layouts like the [Webkit post](https://webkit.org/blog/16026/css-masonry-syntax/) suggests, which will make it easier for developers to learn to new layouts.

## Prefixes and extra properties

One of the main arguments for the Just Use Grid team is the enormous amount of similar properties that would need to be added to CSS.

![image](https://github.com/user-attachments/assets/3090a71c-e04c-4b02-8e25-f4e80a33c114)

By the WebKit Post:

> One of the guiding principles that drives decisions at the CSS Working Group is to always strive to reuse existing patterns and properties when creating new possibilities.

> When multicolumn was invented, a new `column-gap` property provided a way to define the space between columns. A decade later, when the CSSWG needed a way to define space between grid columns, the `column-gap` property was repurposed. It got a corresponding `row-gap` property and a new shorthand, `gap`, to be more universal, and eventually made its way to Flexbox as well. It took a couple of tries, but the CSSWG realized it would be a mistake to keep creating separate gap properties for different contexts — `column-gap` + `grid-gap` + `flex-gap`… It’s better to repurpose what already exists.

By the same logic, it isn't a problem of new properties with a `masonry-`  prefix, but of the prefixes altogether. If we are thinking about defining new layouts, then the `grid-` prefix is also a reusability antipattern. I keep thinking that, just from a syntax standpoint, it should be possible to use a new unprefixed `template-areas` property for masonry and grid, regardless of the formatting contexts.

If we look at another example given in the WebKit post:

![image](https://github.com/user-attachments/assets/a8cdd4cd-191d-4801-a056-d17d3eb7f492)

According to the Chrome proposal, this type of layout would have its own prefix, but I think everyone would agree it would be hellish to also have `pillar-` or `new-layout-`  prefixed properties, and so on if future layouts are added.

Onto the Webkit argument, they say that maybe in the future this layout could be achieved with a snippet like the following:

```css
article {
  display: block; /* default value, unnecessary to state */
  grid-template-columns: 1fr 1fr minmax(15ch, 30ch) minmax(15ch, 30ch) 1fr;
  grid-default-column: 3 / 5;
}
```

But to me, it feels unnatural to use `grid-` prefixed properties for things that aren't grid. That's why I think removing prefixing altogether is an option to consider. The `grid-`  prefixed properties would still work for backward compatibility, but you could also use the unprefixed version. Maybe properties that only work in their corresponding formatting context would remain prefixed, but overall, it would reduce the number of properties to add and the syntax to remember. 

If the conversation is focusing on unifying, then I just wanted to bring this to the table, but I must say that I am ignorant of the consequences and implications of this. Just a not-that-technical developer's point of view of what I would like to code in the future :)

**Comment by tabatkins** [comment]
I've been sitting on this for a bit, because of vacation/Christmas/etc., but I think it's ready to post now.

tldr: The "reuse when possible" principle is *completely valid* and should be (and is) applied as much as possible, but it shouldn't be *overused*, when merging/reuse is messy or complicated. Grid/Masonry is an example of the latter, with several properties that work well in Grid but are *genuinely bad* when used for Masonry, and several other properties that are at least *clumsy* in Masonry (tho technically usable). This appears to be a *general* issue, since Flexbox can't meaningfully merge into these *either*, nor can future examples that have been brought up.

In short, while new layout abilities sometimes merge nicely into an existing layout mode (like Subgrid into Grid), it is *usually* the case that the abilities of one layout mode, and the specialized syntaxes we design to allow authors control over those abilities, are usually a bad fit for even a moderately different layout mode. Layout modes should, in general, get their own set of properties with their own specialized syntaxes, rather than trying to reuse each other's properties.

I'll go into detail on exactly *how* the current Grid-masonry design is bad, and at the end, offer a compromise if the WG still disagrees that Masonry should be its own independent layout mode.

# Specific examples

Apologies for the length of this section. The badness really is in the details; the way that the Grid properties work when applied to Masonry is an *extraordinarily* bad design, and it's worth working thru those details rather than handwaving it as "it's bad, just trust me". If you'd prefer to trust me, feel free to skim.

* [`grid-template-areas`](https://drafts.csswg.org/css-grid-2/#grid-template-areas-property):
	* Syntax allows defining an NxM grid; when used in Grid-masonry, we are forced to ignore all but the first row/column. "Technically valid but ignored" values are an anti-pattern we try to avoid when possible; we often purposely design grammars to avoid this even being possible.
	* A single-column Grid-masonry and a single-row Grid-masonry use different syntaxes; `"one two three"` vs `"one" "two" "three"`. Requiring authors to remember which syntax to use depending on their masonry direction is bad.

		In Grid, the syntax works well, and authors have good reason to break the separate row-strings onto separate lines, to align the columns across the strings and make 2D ASCII art showing the grid structure. There is no such benefit in Grid-masonry (there's only one cell per row), so instead you just have the value put on one line regardless, and the weird syntax split based on direction.

		That is, authors are not going to write:

		```css
		grid-template-areas: "foo bar baz"; // column masonry
		grid-template-areas: "foo"
		                     "bar"
		                     "baz"; // row masonry
		```

		as that's just a frankly silly use of linebreaks/whitespace. They'll just write:

		```css
		grid-template-areas: "foo bar baz"; // column masonry
		grid-template-areas: "foo" "bar" "baz"; // row masonry
		```

* [`grid-auto-flow`](https://drafts.csswg.org/css-grid-2/#grid-auto-flow-property):
	* Grid 2 syntax is `[ row | column ] || dense`. In Grid, this dictates how auto-placed items search for empty spaces - by scanning each row before moving to the next, or scanning each column. Both `row` and `column` *do absolutely nothing* in Grid-masonry; it's already intrinsically auto-placing, and only has one possible axis to do it in. (While `dense` originally did nothing in Masonry either, we've since come up with *a* behavior for it, tho it's less powerful than the Grid behavior.)
	* [Grid 3 adds](https://drafts.csswg.org/css-grid-3/#grid-auto-flow) `row-reverse`, `column-reverse`, and `wrap-reverse`. For Grids, the first two are the same as row/column, but flip the direction you fill the rows/columns in; the third flips to make it fill in the *last* row/column first and wrap back toward the start when you need more rows/columns. 

		In Grid-masonry, we first have the issue that `row-reverse` and `column-reverse` again *do the exact same thing*, because the "auto flow" direction is already fixed. They do finally have *an* effect, tho, which is *vaguely similar* but notably distinct from their effect in Grid: rather than changing the direction you search for holes, it changes which direction you favor when breaking ties for "shortest track". Note that while `row` vs `row-reverse` in Grid will have a pretty large effect on the layout, in Grid-masonry it might *only* have an effect on the first items to be placed (since all tracks are tied initially); afterwards, if there aren't any ties it does nothing.

		Grid-masonry's `wrap-reverse` is more similar to Grid's in effect and impact, as it causes the tracks to fill from their end side. The name is a little misleading, however; there's no real notion of "wrapping" in Masonry (while it makes perfect sense in Grid).

		The `display:masonry` properties are much clearer; they borrow their names and syntax from Flexbox, which has similar concepts: just like Flexbox's `flex-direction: row | column | row-reverse | column-reverse`, we have `masonry-direction: row | column | row-reverse | column-reverse`, *with very similar behavior*. Then the one *novel* property in this group, `masonry-fill`, determines how it breaks ties, a situation/behavior that neither Flexbox nor Grid really has.

* the [`grid-template`](https://drafts.csswg.org/css-grid-2/#explicit-grid-shorthand) shorthand:
	* In Grid, this combines `grid-template-rows`/`grid-template-columns` and `grid-template-areas`. Grid-masonry, thus, inherits all the problems of those properties: only *one* of `-row`/`-columns` is used (the other needs to specify `masonry`), and the `-areas` component has its two distinct syntaxes. The combination actually compounds the badness, tho. 

		First, the `-rows`/`-columns` components are in a *specific* order in Grid: the rows, then the columns. This order is, again, completely sensical in Grid, as it corresponds to how you write the 2d grid in ASCII, but in Grid-masonry it's an unnecessary requirement. So a row-masonry needs to be written `repeat(3, 1fr) / masonry` while a column masonry is `masonry / repeat(3, 1fr)` - the masonry's *direction* is completely implicit, never explicitly mentioned. Compare with `display:masonry`, which uses `masonry: row repeat(3, 1fr)` or `masonry: column repeat(3, 1fr)` (or in reverse order) - direction is clearly and explicitly provided via a keyword, and order of the values doesn't matter.

		Second, the "one string or multiple strings" problem with `grid-template-areas` infects this property too, and compounds. A row masonry is written `"foo" 1fr "bar" 2fr "baz" 3fr / masonry` (with this precisely fixed order of values), while a column masonry is written `"foo bar baz" masonry / 1fr 2fr 3fr`. This *massive* divergence in how you are required to write the property, depending solely on whether it's a row or column masonry, is *completely unacceptable* from a design standpoint. Nobody would *ever* design a property like this on purpose, it's *terrible*.

		Compare with `display:masonry`, which is `masonry: row "foo bar baz" 1fr 2fr 3fr` (with the three bits in *any* order), or `masonry: column "foo bar baz" 1fr 2fr 3fr`. Free re-ordering to whatever the author prefers, and the only difference between the two directions is the keyword *declaring* that direction.

* the [`grid`](https://drafts.csswg.org/css-grid-2/#grid-shorthand) shorthand
	* In Grid, this combines the three properties of `grid-template` with the `grid-auto-*` properties - `grid-auto-rows`, `grid-auto-columns`, and `grid-auto-flow`. One would think that this inherits *even more problems* - all the `grid-template` issues + the `grid-auto-flow` problems, but actually *the property doesn't work at all*.

		That is, while you can of course still write a `grid-template` value in `grid`, it also offers this grammar for doing all both `grid-template-*` and `grid-auto-*`:

		```
		<<'grid-template-rows'>> / [ auto-flow && dense? ] <<'grid-auto-columns'>>?
		```

		(This specifies explicit rows and auto columns; there's also the reverse for auto rows and explicit columns.)

		This grammar *is fundamentally incompatible with Masonry*. It only lets you specify *one of* `grid-template-rows`/`-columns`, but Grid-masonry needs *both* so it can declare one axis with explicit tracks and the other as `masonry`. The most you could do is something like `masonry / auto-flow 1fr`: no explicit tracks listed at all, a meaningless `auto-flow` value (because Masonry is always auto-flowing), and a `grid-auto-columns` value that does get used, but only in a *weird* way: the grid will create enough implicit columns (with that size) to fit the largest spanning item or the furthest placed item. (Which means, in the common case, it'll just spawn a single implicit track.)

		With the Grid 3 values for `grid-auto-flow` it'll be possible to specify a *slightly* meaningful value instead of `auto-flow`, but it'll still mean you only have a single track in the common case. And, again, if you instead wanted to have a row masonry, you'd have to flip the order of things.

		(The grammar *should* be amended to allow specifying `masonry` on the auto-flow side, so the explicit track list is actually useful. But this is another indicator that the mental model underlying the Grid syntax is simply *not designed* for how Masonry wants to be written.)

		Also, the `grid` syntax only lets you *choose* between combining `-rows`/`-columns` with `-areas`, *or* with `-auto-*`. You can't actually shorthand all of the properties at the same time. (This isn't super intentional on Grid's part, fwiw. It would be completely reasonable to want to name your columns even if all your rows are auto-generated. The grammar just gets too complex to make that reasonable, so we left it out.)

		Compare all of this with `display: masonry`, which *trivially* lets you combine all the functionality in a single, fully-reorderable property with no complexity to deal with: `masonry: column-reverse 1fr 2fr 3fr reverse;`.

# Conclusion

Even with a few more fixes to the grammar to make it work properly, Grid-masonry is still *fundamentally* awkward to specify and use. We would *never* design such a syntax from scratch; we'd be laughed out of the WG telcon. The cost of requiring authors to use such a clumsy syntax, and the precedent it establishes for how clumsy of a syntax we're willing to tolerate, is not remotely worth the benefits of authors not having to learn a new set of property names. As well, the benefit of reusing property names does not *meaningfully* carry over to actually reusing the property *values* in many cases; the actual *meaning* of those values is in several cases significantly distinct anyway, so authors are effectively still learning a new set of properties when using Grid-masonry, just spelled the same as an existing set. 

This is actually a *detriment* to learning and understanding, versus having to learn a new-but-similar set of *well-designed* properties and values that match the underlying model better and much less clumsily.

Even if we designed Flexbox from scratch today, it would similarly not be able to reuse the Grid properties without being awkward and clumsy. I assert (without showing my work) that this would be true *even if we designed the entire thing from scratch*, all three of Flexbox, Grid, and Masonry together without any backwards compat to worry about. And the discussed possible future layout models, like "pillar", are similarly *quite distinct* and would have their own awkward/clumsy integration into such a combined layout mode.

In conclusion, Masonry should be its own `display` value, with its own set of properties that are well-designed for it. I *suspect* I (/Chrome) am willing to formally object to the current Grid-masonry design; it really is Just That Bad.

# Compromise (that I don't like) #

If the WG still disagrees with all of the above and thinks it *is* worth trying to reuse Grid properties in Masonry as much as possible, I offer this in-between proposal that avoids as much of the badness as I can.

I must stress that *I do not like this proposal*. A lot of the clumsiness is inherent to the fact that Grid is designed to be fully 2d, while Masonry is, similar to Flexbox, 1d-and-a-bit in its concepts. But it is *substantially* better than the current Grid-masonry proposal.

* `grid-template-rows`/`-columns`/`-areas` reused as-is. Do not add the `masonry` keyword. (Ignore whichever one is in the masonry axis.)
	* Add the "intrinsic sizes in an auto-repeat function" grammar that Masonry allows. Either figure out *some* way to make that usable in Grid, or just give it fallback behavior (take one repetition), like when you use auto-repeats values in an instrincally-sized Grid today.
* `grid-auto-rows`/`-columns` reused as-is. (Again, ignoring whichever one is in the masonry axis.)
* `grid-row`/etc reused as-is. (Again, ignoring whichever one is in the masonry axis.)
* Add a `grid-masonry: none | [ row | column | row-reverse | column-reverse] || [ tie-start | tie-end ]` property, which makes the Grid into a Masonry, and declares the masonry axis and direction, and tie-breaking direction. (Identical to [`masonry-direction` + `masonry-fill` in the current draft](https://drafts.csswg.org/css-grid-3/#masonry-direction), except I'm changing the `masonry-fill` keywords. I want to make that change anyway.) 
* `grid-auto-flow` is ignored entirely; `grid-masonry` completely subsumes it. 
* Add a `masonry` shorthand, [matching the current draft](https://drafts.csswg.org/css-grid-3/#masonry-shorthand), which sets `grid-masonry` and all the existing `grid-*` properties appropriately.

This design lets us ignore the terrible syntax of `grid-template-areas`, `grid-template`, and `grid` (when used for Masonry) entirely; instead, authors will just use the `masonry` shorthand instead, which is simple and easy. `grid-masonry` makes the masonry axis clearer by reusing the already-understandable Flexbox syntax, and lets us avoid the clumsy reinterpretation of `grid-auto-flow` keywords entirely. If you really want to, tho, you *can* still use `grid-template` or `grid` as usual, and then set `grid-masonry` to activate the masonry axis. It does still carry some dead weight; authors have to be aware that several Grid properties are ignored when Masonry is active.

If we *really* wanted to minimize the number of new properties, we could instead merge `grid-masonry` *into* `grid-auto-flow`, as a separate grammar branch with a special keyword, like `grid-auto-flow: masonry row`, and have *that* be the trigger for Masonry behavior. Because this is a completely separate grammar branch, tho, it isn't meaningfully different from just having a separate property. Also, it really does do a *substantially* different thing than what `grid-auto-flow` does today, so merging it in isn't very helpful from a meaningfulness standpoint. I'd recommend against this.

Note that accepting this compromise **does not** imply that we'll continue to merge *more* layout modes into Grid just because they're 2d-ish. We'd still evaluate them as they come.

**Comment by fantasai** [comment]
Apple spent some time reviewing the TAG feedback and the comments above, and particularly considering the concerns about awkwardness with `grid-auto-flow` integration that Google also identified, and we have come up with the following counter-proposal building on the Just Use Grid option:

1. Drop the `masonry` keyword from `grid-template-*`. Continue to use `display: grid`.
2. Re-use the grid templating and grid placement properties, as currently.
3. Unify `grid-auto-flow` with `flex-flow` as `item-flow` per the TAG’s feedback, and split it out into the following longhands as described in https://github.com/w3c/csswg-drafts/issues/11480: `item-direction`, `item-wrap`, and `item-pack`.  Also create `item-slack`, either as a longhand or separate.  (Note: Open to renaming the `item-` prefix.)
4. Add a `collapse` keyword to `item-pack` to trigger masonry-style layout.

The new flow controls for masonry would look like this:

```
item-flow: <item-direction> || <item-wrap> || <item-pack> || <item-slack>
shorthand for...
   item-direction: row | column | row-reverse | column-reverse
   item-wrap: wrap | wrap-reverse
   item-pack: normal | dense || collapse
   item-slack: <length-percentage>
```

<hr>

Here’s the code to accomplish two use cases from [this article](https://webkit.org/blog/16026/css-masonry-syntax/):

![grid-3-layout-light](https://github.com/user-attachments/assets/9fe942af-5251-476c-b919-cb4dd3d24c2f)

```
.masonry {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(14rem, 1fr));
  item-flow: collapse; /* == item-flow: row collapse 1em; */
  gap: 1rem; 
}
```

![sideways-masonry-light](https://github.com/user-attachments/assets/77963065-7c1d-4a8a-8953-545161a0369c)

```
main {
  display: grid;
  grid-template-rows: 2fr 1fr;
  item-flow: column collapse; 
}
```

Thoughts?

**Comment by Crissov** [comment]
`item-` → `tile-`

**Comment by Loirooriol** [comment]
> Thoughts?

IMO it looks quite confusing that in order to get a masonry that has rows, you need `item-flow: column collapse`. Sure, items will flow vertically so you can think of columns, but there are no actual columns. And I think it's much more intuitive to define whether you want columns or rows, rather than defining whether you want the next item to flow vertically or horizontally, and then infer from there. So I still prefer `masonry-direction` & friends.

**Comment by JoshTumath** [comment]
I commented on Apple's proposal: https://github.com/w3c/csswg-drafts/issues/11480#issuecomment-2590784323

I like it, but I'm wondering if it leads us towards a hybrid of Apple and Google's syntax proposals. Each layout system still has properties unique to it, like `flex` or `grid-template-columns`. Where it's not possible to unify layout properties as per the TAG review recommendation, maybe it does make sense, after all, to have `masonry-*` properties?

So I'm interested in what other opportunities we have to unify the syntax, which may help to address Tab's concerns. (https://github.com/w3c/csswg-drafts/issues/11243#issuecomment-2563983858).

**Comment by tabatkins** [comment]
(I commented over in #11480 instead.)

**Comment by kbabbitt** [comment]
Thinking about the TAG's question of what we would have done if we had a time machine... I might have proposed a refactoring of Grid+Flexbox into:
1. A scheme for specifying layout boundaries in a 2D space, in both horizontal and vertical directions. I'm reminded of "guidelines" from the desktop publishing world. In CSS, we have something similar in track definitions, except we specify the spaces between the guidelines rather than the guidelines themselves. Such a scheme could be shared across various 2D layouts, with some using both sets of guidelines/tracks and some using just one.
2. A scheme or schemes for specifying how children are arranged within that space. This is how we would define wrapping-flexbox, grid, masonry, Flickr, and maybe even other 2D layouts such as Pillar. The `item-*` properties and discussion in [#11480](https://github.com/w3c/csswg-drafts/issues/11480) seem to me like a great step towards a common language for these.


To those points, I think this: 
> Drop the masonry keyword from `grid-template-*`

is an important simplifying change, because it decouples track definition (point 1; what existing `grid-template-*` were doing) from item placement (point 2; what existing `grid-auto-flow` was doing).

**Comment by jfkthame** [comment]
>     1. A scheme for specifying layout boundaries in a 2D space, in both horizontal and vertical directions. I'm reminded of "guidelines" from the desktop publishing world.

As something of an aside, perhaps this is also how the features of "rhythmic sizing" or "line-grid" should be introduced. They're just more cases of defining a "grid" (in one direction), and then controlling how elements size and place themselves in relation to that grid.

**Comment by juanrgon** [comment]
I’m trying to fully understand the resolution. Which of these is correct:

1. Masonry will be part of `display: grid` and will not introduce new `masonry-` properties?
2. _If masonry_ were to be built into `display: grid`, it would reuse Grid properties rather than adding new `masonry-` props?
3. Regardless of whether masonry is implemented within `display: grid` or as `display: masonry`, it will still reuse existing grid properties?

**Comment by tabatkins** [comment]
Those minutes are part of a multi-part discussion that's still ongoing, so they're not answered yet. ^_^

**Comment by tabatkins** [comment]
Ah, sorry, thought those were three questions. Its (3); regardless of how we trigger masonry layout, we'll use the grid-template-* and grid placement properties on it.

**Comment by kbabbitt** [comment]
I think I did a poor job of articulating my position on the Masonry Layout Switch in the room today, so I'd like to try again.

What I like about the `item-*` properties, as an idea, is that they factor out concepts that make sense, and can represent the same concept, across multiple layout types: whether a 2d layout is arranged principally in rows or columns; the wrapping behavior for a 2d layout; the amount of "wiggle room" when placing items in a 2d layout. That lets me mentally group these properties into the same bucket as other property families that work across these contexts: gap, rule, even backgrounds and borders.

I feel the same way about grid-template-rows and grid-template-columns with the `masonry` keyword dropped from the syntax. Given a time machine I might rename them to something without the word "grid" to indicate that tracks/guidelines are a more general concept; I don't know that aliasing them now for that reason would be high value. The point is that tracks/guidelines make sense across multiple 2d layout types.

`item-pack: collapse` as a grid/masonry switch isn't quite there yet for me. Thinking about it from a greenfield design scenario - or through the eyes of someone learning these for the first time - I think "collapse tracks of a 2d container in one direction" could be interpreted either as masonry or as wrapping-flexbox. There's a missing ingredient which is *how the descendant items are laid out* - which, as @alisonmaher pointed out, is exactly the inner display type expressed by [the `display` property](https://drafts.csswg.org/css-display/#the-display-properties) - and it's how we currently differentiate between grid and wrapping-flexbox. Perhaps there's an `item-*` which could supply that information in a consistent way. But with what we have so far, that's why I still prefer `display: masonry` to complement the great work that's been done so far to refine the proposals.

**Comment by schenney-chromium** [comment]
I really appreciated the thorough discussion at the F2F and it helped a lot. I support the item-blah properties and would wait until we sort that out before deciding finally on the switch.

After listening and thinking, there doesn't seem to be any strong reason to favor one approach over the other (`display` vs property). Each side has reasonable arguments, I don't think the developer ergonomics are different enough to force a decision. So how to come to agreement after many years of discussion?

Browser performance and binary size comes to my mind. Could implementors comment on how often they would write `if (display == grid || display == masonry) ` in layout code vs `if (someProperty == doMasonry)` in layout code? If the former, in which more code is common between masonry and grid, then make the property the switch. If the latter, where more code differs between layout and grid, then use display as the switch.

Sorry if this has been considered already, but it did not come up in the latest discussion.

**Comment by JoshTumath** [comment]
So one of the key issues is how to set the tracks for masonry layout. Expanding on the `item-*` properties idea, perhaps there's a way we can make that work for track/column/row templates as well... (Apologies if I'm re-stating an idea that someone's already said.)

### Problem

So I really like the `item-*` properties. @fantasai gave this example:

```css
.rows {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(14rem, 1fr));
  item-flow: row collapse;
}

.columns {
  display: grid;
  grid-template-rows: 2fr 1fr;
  item-flow: column collapse; 
}
```

But in a masonry layout mode, what would happen if I tried to set the template for the cross axis? For example...

```css
.rows {
  display: grid;
  grid-template-rows: 2fr 1fr; /* Would this work? The row axis is collapsed */
  item-flow: row collapse;
}

.columns {
  display: grid;
  grid-template-columns: 2fr 1fr; /* Would this work? The column axis is collapsed */
  item-flow: column collapse; 
}
```

It feels very weird to me that we're using `item-flow`/`item-direction` to specify which axis collapses, but, at the same time, when we want to size the tracks, we still have to pick the property that will work: `grid-template-columns` or `grid-template-rows`. (I think similar to one of @tabatkins's criticisms of the 'just use grid' proposal.)

### Proposal

**I'm proposing two new `item-*` properties: `item-template-main-axis` and `item-template-cross-axis`**.

Think of them as the grid/masonry equivalent of what logical properties are to physical properties, so they'd co-exist with `grid-template-*` in the same way as `margin-block-start` and `margin-top`.

So the above example could be rewritten like so:

```css
.masonry {
  display: grid;
  item-template-main-axis: 2fr 1fr;
  
  &.row {
    item-flow: row collapse;
  }
  
  &.column {
    item-flow: column collapse;
  }
}
```

How the properties would work in each layout mode:

* In the collapsed/masonry mode, the `item-template-cross-axis` property gets ignored. 
* In the grid mode, `item-template-cross-axis` property works as you would expect.
* In the flex layout mode, both the `item-template-main-axis` and `item-template-cross-axis` properties would presumably be ignored.

By the way, enabling masonry layouts via `display: masonry` would still work under this proposal.

> [!NOTE]
> I'm not proposing an `item-template-area` property. In my opinion, the concept of 'template areas' probably doesn't make sense for masonry and it should remain specific to grid layouts.
>
> But I note that [the original `masonry-*` properties proposal did include `masonry-template-tracks`.](https://www.w3.org/TR/2024/WD-css-grid-3-20241003/#masonry-track-masonry-option) I know we agreed not to use the `masonry-*` properties, but perhaps it does make sense to have `masonry-*` property for that specific purpose after all. Or maybe there is some way of making it work in a layout-agnostic `item-*` property?

So what do you all think? Could this work? Does it make sense?

**Comment by tabatkins** [comment]
The template properties weren't part of the item-* set because they didn't seem quite layout-agnostic. Flexbox doesn't have such a notion, for example, and even if it does grow one (to help support balancing), there's no guarantee it will look similar to the grid/masonry properties. 

There are also additional concept they mix in, like grid-template-areas, which also aren't necessarily useful across all layout modes. Shorthands, also, lean toward wanting to be specific to the layout mode. (I've argued strongly for this; that `grid-template` and `grid`, while fine for Grid, are *terrible* for Masonry.)

>  in a masonry layout mode, what would happen if I tried to set the template for the cross axis? For example...

It's ignored. The template and placement properties only work if they're in the grid axis of the masonry.

While there are some arguable downsides to this (as you point out, I've given those arguments before), there's also benefits, in that the sizing axis isn't indirected thru another property, which requires you to think about what axis you're specifying. For example, I think on balance it was a mistake to have `flex-basis` (we should have just used `width`/`height`) and to have the justify/align properties on Flexbox pay attention to main/cross axis (they should just be inline/block axis, like in all other layout modes). I think that it's often easier for authors if properties use a consistent axis (at least, consistent within a language, like the logical directions; authors rarely mix language directions on a page so it doesn't usually require extra thought).

**Comment by fantasai** [comment]
Summarizing where we are:
- CSSWG [resolved](https://github.com/w3c/csswg-drafts/issues/11243#issuecomment-2627998471) to re-use the `grid-*` properties for templating and explicit placement. This is now edited into the draft.
- CSSWG resolved, in principle, to pursue the [`item-flow` proposal originally suggested here](https://github.com/w3c/csswg-drafts/issues/11243#issuecomment-2584991716) and described further in https://github.com/w3c/csswg-drafts/issues/11480. The exact syntax is still being debated there.
- How we switch into masonry layout is still an open question. https://github.com/w3c/csswg-drafts/issues/12022
- Masonry-related shorthand additions and adjustments are still an open question. Filed https://github.com/w3c/csswg-drafts/issues/12023 for follow-up.
- https://github.com/w3c/csswg-drafts/issues/11593 serves as a table of contents for these debates.

I think we should probably close this issue and continue in the individual follow-ups?
